flocker/acceptance/test_restarts.py

```diff
@@ -0,0 +1,169 @@
+            created = cluster.create_container(http_server)
+            created.addCallback(lambda _: self.addCleanup(
+                cluster.remove_container, http_server[u"name"]))
+            created.addCallback(lambda _:
+                                loop_until(lambda: query_server().addErrback(
+                                    lambda _: False)))
+            created.addCallback(lambda _: query_server())
+            return created
+        creating_dataset.addCallback(created_dataset)
+
+        def server_started(initial_response):
+            # We now have the initial response of the server. This should
+            # be the initial boot time of the node, repeated twice.  Next,
+            # we restart the machine.
+            print "INITIAL RESPONSE", initial_response
+            initial_reboot_time = initial_response.splitlines()[0]
+            initial_container_id = initial_response.splitlines()[2].encode("ascii")
+            # XXX Checking exit code is problematic insofar as reboot
+
+            # kills the ssh process...
+            print "Rebooting!"
+            disabling = _service(node.public_address, b'flocker-dataset-agent', b'disable')
+            def reboot(ignored):
+                call([b"ssh", b"root@{}".format(node.public_address),
+                      b"shutdown", b"-r", b"now"])
+            rebooting = disabling.addCallback(reboot)
+
+            # Now, keep trying to get a response until it's different than
+            # the one we originally got, thus indicating a reboot:
+            def query_until_different():
+                print "query!"
+                return query_server().addCallbacks(
+                    lambda response: response != initial_response,
+                    lambda _: False)
+
+            # Now, poll the host until the now-stopped Docker container that existed
+            # on the first run is destroyed. This is an external sign that the container
+            # agent tried to "restart" (by destroying and recreating) the container.
+            # If the bug that this test is designed to catch occurs, then the container
+            # will have been stopped and a new one started before the dataset is in place.
+            # The dataset cannot be in place because the dataset agent isn't running.
+            # So by the time (just after) the container agent kills the first container
+            # ID, the bad thing will already have happened, and we can start the dataset
+            # agent in order to allow a correct implementation (where the bug is avoided)
+            # to eventually get the dataset in place and start the container correctly.
+            def old_container_gone(ignored):
+                if call([b"ssh", b"root@{}".format(node.public_address),
+                         b"true"]) == 0:
+                    if call([b"ssh", b"root@{}".format(node.public_address),
+                             b"docker", b"inspect", initial_container_id]) == 0:
```

Comment:
The docker HTTP API is directly accessible, so we can use docker-py and the Docker API here.